### ISSUE TITLE
Design Systems VNext: Design Tokens

### ISSUE THREAD

**Issue by nicholasrice** [microsoft/fast#4369]
With the addition of Dependency Injection and new component registration methods provided by `FoundationElement` and `DesignSystem`, this issue proposes a new way to expose Design Tokens to an application using those features.

**Goals:**
- Make Design Tokens a top-level API surface to better reflect industry standards and terminology
- Remove the need of a wrapping `fast-designs-system-provider` element
- Provide mechanisms to make Design Token implementations injectable and customizable by app authors / component libraries.
- Provide mechanisms to allow Design Tokens to have complex hierarchies.
- Simplify difficulties in setting the `background-color` design token to a color recipe implementation.
# Design Tokens
Design tokens are a fundamental building block for creating design systems, and inform individual property values from the design language. Examples of design tokens are colors, font sizes, spacing units, etc.

## Requirements
1. Accessible through DOM tree
2. Can have values reflected to CSS custom properties at arbitrary DOM nodes.
3. Mutable through the DOM tree. A token value at one node may not be the value at another node.
4. Observable. It must be possible to know when a token changes.

## Implementation of a Design Token
A design token is a typed, unique symbol that can be used to resolve *values*. The generic parameter identifies the runtime type the token will resolve.

```ts
type DesignToken<T> = {};
```

## Design Token Library
The token library is responsible for *managing* design tokens. It is where token values are set and retrieved, and is responsible for emitting notifications when tokens are set. 

I was originally hoping the DI Container could be used to manage setting and getting values of Design Tokens but ran into the following issues:
1. We need to optionally reflect a token to CSS custom properties when it is set. This isn't a behavior of DI containers nor should it be, so we need *some* abstraction here.
2. DI containers don't *re-resolve* if an intermediary Container registers a value. This becomes problematic when a Design Token is set during an app lifecycle at a node it was not previously set; this is invisible to previously resolved values.

Because of these limitations and requirements, a purpose-specific resolution and notification system will be required.

**DesignTokenLibrary Interface**
```ts
interface DesignTokenLibrary {
    /**
     * Enables hierarchial properties through a tree
     */
    upstream: DesignTokenLibrary | null;
   
    subscribe(subscriber: Subscriber, ...tokens: DesignToken<any>[]);
    unsubscribe(subscriber: Subscriber, ...tokens: DesignToken<any>[]);
    get<T>(token: DesignToken<T>): T;
    alias<T>(alias: DesignToken<T>, target: DesignToken<T>): void;

    set<T>(token: DesignToken<T>, value: T): void;
    set<T>(token: DesignToken<T>, value: DerivedDesignTokenValue<T>): void;

    static createToken<T>(): DesignToken<T>;
}
```

## Design Token Values
Any Design Token can be assigned a static value or `DerivedDesignTokenValue` through the `DesignTokenLibrary`.

### Setting Static Values
A static value has no dependencies or calculating logic; it is just a value.

```ts
const DesignUnit: DesignToken<number> = DesignTokenLibrary.createToken<number>();

const library = new DesignTokenLibrary();
library.set(DesignUnit, 4)
```

### Setting Derived Token Value
Derived Design Token values are values that have dependencies and calculating logic. The value itself is responsible for deriving the *derived* value.

```ts
interface DerivedDesignTokenValue<T> {
    dependencies: Array<DesignToken<any>>;
    derive(dependencies: Array<any> /* The resolved dependencies */): T;
}

const ButtonHeight: DesignToken<number> = DesignTokenLibrary.createToken<number>();
const value: DerivedDesignTokenValue<number> = {
    dependencies: [DesignUnit],
    derive: (args: [number]) => {
        const [ designUnit ] = args;
        return designUnit * 4;
    }
}

library.set(ButtonHeight, value);
```

## The DesignTokenProvider
A `DesignTokenProvider` is responsible for creating a new token library for the element instance, setting the upstream library based on DOM location, and reflecting properties set on the instance to CSS custom properties on the element.

```ts
interface DesignTokenProvider extends FASTElement {
    /**
     * Library instance
     */
    designTokenLibrary: DesignTokenLibrary;

    /**
     * Emits the Design Token as a CSS custom property
     */
    toCSS(token: DesignToken<T>): string; // CSS custom property name
}
```

The provider manifests as a mixin class, and creating a provider looks like:

```ts
class MyProviderElement extends DesignTokenProvider(FoundationElement) {}
```

## Initializing and Configuring Design Tokens
Authors can use the DesignSystem to configure a Design Token, setting an optional default value, an optional CSS custom property name to use, and optional formatter functions to use when emitting the Design Token to a CSS custom property.

```ts
interface DesignTokenConfiguration<T> {
    token: DesignToken<T>,
    value?: T | DerivedDesignToken<T>,
    customPropertyName?: string,
    formatter?: [(value: unknown) => any]
}

DesignSystem.configureDesignToken(configuration: DesignTokenConfiguration);
```

The DesignSystem itself will register a `DesignTokenLibrary` that reflects CSS custom properties to the document root. This will allow us to remove the "root provider" requirement that we currently have

```html
<!-- We don't need this anymore -->
<fast-design-system-provider use-defaults>

</fast-design-system-provider>
```

## Contextual Color System 
With all of the above, color behaviors can be written in dynamic ways to facilitate a variety of scenarios. The implementation of the `NeutralForegroundRest` design token can be separated from the token itself and resolved through Dependency Injection, providing opportunity to customize color recipe implementations.

```ts
const NeutralForegroundRestRecipe = DI.createInterface<DerivedDesignTokenValue<string>>(x => x.instance({
    dependencies: [/*...*/],
    derive: () => { /* some implementation */}
}));

DesignSystem.register(NeutralForegroundRestRecipe);

// Later
class NeutralForegroundRestBehavior {
    bind(element) {
        const container = DI.getOrCreateDOMContainer(element);
        const tokenProvider = container.get(DesignTokenProvider);

        if (!tokenProvider.has(NeutralForegroundRest)) {
            const value = container.get(NeutralForegroundRestRecipe)
            tokenProvider.set(NeutralForegroundRest, value)
        }

        tokenProvider.toCSS(NeutralForegroundRestRecipe);
    }
}

const style = css`
    :host {
        color: ${NeutralForegroundRestBehavior()}
    }
`
```

Additionally, setting the "background color" Design Token to a color Recipe becomes very simple:

```ts
class MyCard extends DesignTokenProvider(FoundationElement) {
  connectedCallback() {
    const container = DI.getOrCreateDOMContainer(this);
    this.designTokenLibarary.set(BackgroundColor, container.get(NeutralFillCardRecipe))
  }
}

```

**Comment by nicholasrice** [comment]
Additionally, scaffolding the current implementation of the DesignSystemProvider element with these APIs becomes very easy:

```ts
class FastDesignSystemProvider extends DesignTokenLibrary(FoundationElement) {
  @attr
   designUnit: number;
   designUnitOnChanged() {
      this.designTokenLibrary.set(DesignUnit, this.designUnit);
   }
   // ...
}
```

**Comment by EisenbergEffect** [comment]
Ok, I'm super duper excited about this. So much so that I've been thinking a ton about it for the last week. I think we're on an excellent path here. I've got a bunch of feedback which will hopefully be useful. Including some pseudo-code to help with some of the ideas. Some of it has obvious holes so @nicholasrice would certainly need to tell me if it wouldn't work. In general, it focuses on expanding the API of what a token is, and trying to take advantage of some existing infrastructure in fast-element that is probably less well-known in terms of how to use it directly.

First, everything below assumes that we've added CSS directives to fast-element. So, the `css` template function would recognize values inherited from this type:

```ts
export abstract class CSSDirective {
  createCSS(): string {
    return '';
  }

  createBehavior(): Behavior | undefined {
    return undefined;
  }
}
```

Ok, my first thought was, can we take this idea of a design token and make it more than just a key? When I read through the proposal, I later saw the design token configuration and a few other things in different places. I wondered if we could/should aggregate a few things together as part of the token itself. So, given a token, you could know its name, see if it writes css variables, get its value for a particular element, set its value for an element, and also use it directly in CSS. So, something like this:

```ts
export type ComputedTokenInput<T> = (element: HTMLElement) => T;
export type TokenInput<T> = T | ComputedTokenInput<T>;

export class DesignToken<T = any> extends CSSDirective {
  private constructor(public readonly string: name, public readonly writeCSSProperty = true) {}

  public getValue(element: HTMLElement): T {
    return TokenValue.find(element, this).value;
  }

  public setValue(element: HTMLElement, input: TokenInput<T>) {
    TokenValue.find(element, this).value = input;
  }

  createCSS(): string {
    return `var(--${this.name})`;
  }

  public static create<T>(name: string, writesCSSProperty = true): DesignToken<T> {
    // check if name already exists globally and throw if it does
    const token = new DesignToken(name, writesCSSProperty);
    // register token by name
    return token;
  }
}
```

So, the idea here is that we keep the token metadata, such as whether it writes a css property, along with its name, as part of the token itself, we can then also use the token directly to set/get its value without needing a central library API. So, parts of the library would be implemented as part of token, and parts might be implemented as part of this other thing I called `TokenValue` (more on that below).

So, a few things you could do with this....

Define tokens (very similar to above):

```ts
export const background = DesignToken.create('background');
export const foreground = DesignToken.create('foreground');
``` 

Use these tokens in styles:

```ts
const styles = css`
  :host {
    background: ${background};
    color: ${foreground};
  }
`;
```

Add some APIs to `DesignSystem` so that they can be set to default values for a particular system implementation:

```ts
export function fastDesignSystem() {
  return new DesignSystem()
    .withTokenMap({
      [background.name]: 'black',
      [foreground.name]: ele => {
        const backgroundValue = background.getValue(ele);
        return backgroundValue * doSomeMath;
      }
    });
}
```

Override the tokens when configuring a design system for use in an app:

```ts
fastDesignSystem()
  .withToken(foreground, 'cornsilk') // override a token value
  .register(
    button(),
    select(),
    card()
  ).applyTo(document);
```

Arbitrarily set a token on an element:

```ts
const element = ...get an element somehow...
foreground.setValue(element, 'yellow');
```

We could even expand the token to let you *set* its value in CSS through the CSS behaviors:

```ts
const styles = css`
  ${background.override('blue')}
`;
```

The `override` method would return a `CSSDirective` that produced a`Behavior` to set the token value on the element. Not particularly useful for our component library, but perhaps very useful for experience and app authors. So, I'm thinking of this as a variation on the `DesignTokenProvider` that allows us to maybe get away from requiring an intermediary element. Of course, you could always have that as well.

So, hopefully this makes sense so far. I was trying to see if we could make the token itself more interesting and remove the need for everything to be done through special config or library APIs. I think a good bit of the code would be similar, but just moved around.

Ok, the second thought I had was about this thing I called `TokenValue`. Looking at the library abstraction and the derived token API, I kept thinking that it was very similar to what Observables can do in fast-element. We can subscribe, and unsubscribe. We can also have computed expressions that are observed. What if we took this idea of a *value* and separated it from the token, made it observable itself, and used our existing infrastructure to handle subscribe/notify and computed dependencies. With that in mind, here's some pseudo code based on lower level fast primitives...

```ts
export class TokenValue extends SubscriberSet implements Subscriber {
  private $fastController = this; // HACK (we can fix this in fast-element)
  private _value: any;

  constructor(public readonly token: DesignToken, private readonly element: HTMLElement) {
    super(this); // should this be the token instead of this?
    // register for DOM events on the element for this token; similar to DI
  }

  get value() {
    Observable.track(this, 'value');
    return this._value;
  }

  set value(value: any) {
    if (typeof value === 'function') {
      this.observer = Observable.binding(value, this); // TODO: get or create
      this.writeValue(
        this.observer.observe(this.element, defaultExecutionContext)
      );
    } else {
      this.writeValue(value);
    }
  }

  handleChange(source: any, args: any): void {
    // one of the dependencies of this token value has changed
    this.writeValue(
      this.observer.observe(this.element, defaultExecutionContext)
    );
  }

  writeValue(value) {
    this._value = value;

    if (this.token.writeCSSProperty) {
      // ...
    }

    this.notify('value');
  }

  public static find(element: HTMLElement, token: DesignToken) {
    // raise event which will be handled by an token value instance
  }
}
```

First thing to note is that `TokenValue` inherits from `SubscriberSet`. This is a fast-element API that implements the `Notifier` API. Basically, you can subscribe, unsubscribe, and notify. So, now we have those basic capabilities for a single value. Next, we make it implement `Subscriber`. This enables it to subscribe to its dependencies, if its value is a function that computes the derived value. Internally, the `value` calls our `Observable` APIs for dependency tracking and value change notification. This means that we can use `Observable.binding` to observe an arbitrary function containing design tokens and observables, producing a token value and notifying any subscribers of its own. The final piece is to steal a trick from @nicholasrice's DI work for parent/child containers. When a token value is set on an element, we can `addEventListener` to listen for any requests to find the token value. The publish of the event would be abstracted behind the static `find` method.

Again, a lot of this was me sitting down and hacking to find out if we could somehow build this on the low level primitives of Observables already in fast-element, integrating those systems. Hopefully you can make some sense out of this.

**Comment by nicholasrice** [comment]
This looks awesome @EisenbergEffect! I like the idea of giving the `DesignToken` a bigger role, making it usable in stylesheets directly, and associating some metadata with it. One aspect I'm un-sure of is how best to support the "asynchronous intermediary" scenario, let me try to explain:

1. Three elements, *A*, *B*  and *C* exist, where *B* is a descendent of *A* and *C* is a descendent of *B* (*A* -> *B* -> *C*) .
2. *A* sets token *_a* on itself
3. *C* asks for token *_a* and it is resolved from *A*.
4. *B* sets token *_a* on itself sometime later
5. -> This is where *C*'s *_a* should change, but it's observing *A* not *B*. I think we'll need a mechanism to track who is asking for what and teardown / build up new subscribers right? The same type of problem exists for dependencies of a `ComputedTokenInput` I think.

The above scenario was the primary reason I proposed a `DesginTokenLibrary` where it would be easier to track changes to values where they had previously been unset (as well as deletions), but maybe (hopefully) I'm missing a simple way this could work with observable?

**Comment by EisenbergEffect** [comment]
That all makes sense. I also realized there was a challenge around computed tokens being based on the wrong inputs if the inputs were overridden at a child node.

My thought was that the interesting operation is the operation of setting a token value. So, it may be that any time a token is set on a node, we need to pull down all tokens above and then invalidate computed tokens below. I think that something like your `DesignTokenLibrary` would definitely be needed to manage that. I wonder if it's less of a library and more of a storage mechanism. Also, I think that it should be treated as more of an implementation detail behind the scenes as opposed to something that people interact with directly. So, devs still just create tokens, set them, and use them with an API like I described above, but behind the scenes there is the `TokenValue` and maybe some sort of `TokenStorage` that manage the details of subscription and invalidating computeds, etc. 

Thoughts?

**Comment by nicholasrice** [comment]
Yea I think that will work; I'll do some explorations and see how we can do that efficiently.

**Comment by KingOfTac** [comment]
From an outsider's perspective, I think this will be huge for component authors. You can definitely count me among the early adopters of these features.

**Comment by bheston** [comment]
I agree with the overall evolution of the design and functionality here.

I do believe the intermediary issue described is something we'll need to address or it will lead to confusion during implementation.

I also really like the scaffolded `FastDesignSystemProvider` example. That seems like the perfect typed wrapper to the generic foundation. It essentially becomes the definition of any particular design system, in the sense that I think @chrisdholt often refers to it. That is, not the function of the components, but all the stylistic values that make something FAST, Fluent, Material, custom, etc. It would be as complicated as needed to express the visual design. I think it would still be beneficial to have some sort of interface to this so we can maintain some common attributes across systems. For instance, density is something that keeps coming up and is supported by many design systems. There may be no standard for it now, but it's conceivable you could have FAST, Fluent, Material, etc. design systems that implemented a common density model to allow for potential runtime switching across various attributes (compact FAST, spacious Material, normal Fluent, etc.)

Coupled with #4381 I think this provides the modular approach we've been discussing.

One thing I'd still like to find a way to address is around handling _sets_ of tokens. The best example of this in our current design system model is color states. Those commonly have a backing algorithm for the set, but we treat each state as an individual "recipe". This seems like something we could consolidate similar to @EisenbergEffect's suggestions for the various layers of design token configuration here. From prior debugging, I believe our current model causes the algorithm to be invoked up to three times (for rest, hover, and active states) and it creates a lot of boilerplate code for the `SwatchRecipe`s. As it relates to this issue, I think those would still be exposed as three design tokens, but somehow the set would tie together the `neutral-foreground` portion and add `rest`, `hover`, and `active` specifics.

The contradiction to this example is some extreme composition where someone wants to use the `neutral-foreground-rest`, `accent-foreground-hover`, and `fancy-foreground-active` recipes for a single style, but perhaps that could be addressed as a composite support in #4381.

One other thought is, I wonder if there can be a simpler way for setting a component's background color as in your example. I suspect it might be simplified considering your subsequent comment regarding the `FastDesignSystemProvider`, but I have a feeling this ties back to my earlier comment about some sort of base design system. I believe that if `MyCard` extends some Foundation class that there would be no opportunity for it to simply have a `backgroundColor` attribute that could easily and uniformly be set. Regardless I think it's important to be able to set this externally and not just on the connected callback. Perhaps `<my-card background-color='#F0F'>`.

**Comment by nicholasrice** [comment]
I think we could use the existing `Behavior` / `CSSDirective` infrastructure to create recipe sets, where binding the behavior sets up hover / click /etc listeners and sets the `background-color` design token to the appropriate recipe:
```ts 
const myRecipeSet = new ElementStateFillBehavior(restRecipeToken, hoverRecipeToken, activeRecipeToken, focusTokenRecipe) // Or somthing
```

The component would use the recipe set and set the background color:

```ts
css`
   ${myRecipeSet} /* This could also happeny anywhere else and could be injected by the DesignSystem w/ #4381 */
    :host {
        background-color: ${backgroundColorDesignToken}
    }
`
````

The above isn't limited to fills either, and we can figure out how best to organize design concerns into asset packages that make sense for consumers and core scenarios.

Setting a background color for a component is pretty easy in this model:

```ts
const backgroundColor = DesignToken.create<Color>("background-color");

// later
backgroundColor.setValueFor(targetFASTElement, "#F0F");
```

The above operation can happen externally or can be set by the target element itself:
```ts
backgroundColor.setValueFor(this, "#F0F")
```
